### PR TITLE
api: stream data in api.open() for S3, GS and OSS

### PR DESCRIPTION
**Pull request by Suor**
Employ the same sign url and use resumable download approach tested for
Azure.

### CODE REVIEW

File: dvc/remote/base.py [hunk 1]
```diff
@@ -515,6 +516,10 @@ def download(
         return 0
 
     def open(self, path_info, mode="r", encoding=None):
+        if hasattr(self, "_generate_download_url"):
+            url = lambda: self._generate_download_url(path_info)  # noqa: E731
```
Comment (efiop) [inline dvc/remote/base.py]: Let's use def as PEP-8 recommends. https://www.python.org/dev/peps/pep-0008/#programming-recommendations

**Comment by efiop** [comment]
Btw, please rebase :slightly_smiling_face:

**Comment by efiop** [comment]
Really glad to see it all got generalized so nicely for s3,gs,azure,oss! Looks amazing!

**Comment by Suor** [comment]
Rebased and removed `lambda`.

**Pull request merged**